lib/penelope/nlp/tokenize/bert_tokenizer.ex

```diff
@@ -0,0 +1,155 @@
+    inspect(@defaults.lowercase)
+  }``|
+  |`split_regex`|regex used to tokenize the text|``#{
+    inspect(@defaults.split_regex)
+  }``|
+  |`strip_regex`|regex used to remove invalid characters|``#{
+    inspect(@defaults.strip_regex)
+  }``|
+  |`punct_regex`|regex used to split pieces on punctuation|``#{
+    String.replace(inspect(@defaults.punct_regex), "|", "\\|")
+  }``|
+  |`special_tokens`|list of special tokens not to piece|``#{
+    inspect(@defaults.special_tokens)
+  }``|
+  |`unknown_token`|the key used to indicate OOV token|``#{
+    inspect(@defaults.unknown_token)
+  }``|
+  |`piece_prefix`|prefix used to indicate subsequent pieces|``#{
+    inspect(@defaults.piece_prefix)
+  }``|
+  """
+  @spec encode(
+          text :: String.t(),
+          vocab :: %{required(String.t()) => integer()},
+          options :: keyword()
+        ) :: {[String.t()], [integer()], [integer()]}
+  def encode(text, vocab, options \\ []) do
+    # apply customizations over the default tokenizer configuration
+    config =
+      options
+      |> Enum.into(@defaults)
+      |> Map.put(:vocab, vocab)
+      |> Map.update!(:special_tokens, &MapSet.new/1)
+
+    # the base tokenization is simply to split on whitespace
+    tokens = String.split(text, config.split_regex, trim: true)
+
+    # convert the tokens to vocabulary keys
+    # tokens can be split by punctuation and further split into
+    # pieces, so we maintain an index into the original token list
+    {indexes, keys} =
+      tokens
+      |> Enum.with_index()
+      |> Enum.flat_map(&split(config, &1))
+      |> Enum.flat_map(&piece(config, &1))
+      |> Enum.unzip()
+
+    # vectorize the keys through the vocabulary
+    values = Enum.map(keys, &config.vocab[&1])
+
+    # return the tokens, piece indexes, and value vector
+    {tokens, indexes, values}
+  end
+
+  @doc """
+  detokenizes a (possibly sub-)sequence of an encoded string
+  """
+  @spec decode({[String.t()], [integer()], [integer()]}) :: String.t()
+  def decode({tokens, indexes, _values}) do
+    # to convert back to the original text, retrieve the
+    # distinct list of tokens for each requested piece,
+    # and then join on spaces
+    indexes
+    |> Enum.dedup()
+    |> Enum.map(&Enum.at(tokens, &1))
+    |> Enum.join(" ")
+  end
+
+  defp split(config, {token, index}) do
+    if token in config.special_tokens do
+      # don't split reserved tokens ([PAD], etc.) into pieces,
+      # so they can be vectorized as-is
+      [{token, index}]
+    else
+      # normalize the token and split it on punctuation
+      token
+      |> String.normalize(:nfd)
+      |> (&if(config.lowercase, do: String.downcase(&1), else: &1)).()
+      |> String.replace(config.strip_regex, "")
+      |> String.split(config.punct_regex, include_captures: true, trim: true)
+      |> Enum.map(&{&1, index})
+    end
+  end
+
+  defp piece(config, token) do
+    piece(config, _pieces = [], token, _prefix = "")
```

Comment:
If you were so inclined, you could avoid dealing with the tuples in the helper fns (since it's not needed there) by pattern matching it out in this signature and mapping it onto the results of the call to `piece/3`, but I suppose that'd make things one iteration slower.

Reply:
that's a good idea, will do